lib/ansible/modules/cloud/amazon/iam_policy.py

```diff
@@ -112,240 +114,215 @@
-        updated_policies = [cp for cp in iam.get_all_user_policies(name).
-                            list_user_policies_result.
-                            policy_names]
-    except boto.exception.BotoServerError as err:
-        error_msg = boto_exception(err)
-        module.fail_json(changed=changed, msg=error_msg)
-
-    return changed, name, updated_policies
-
-
-def role_action(module, iam, name, policy_name, skip, pdoc, state):
-    policy_match = False
-    changed = False
-    try:
-        current_policies = [cp for cp in iam.list_role_policies(name).
-                            list_role_policies_result.
-                            policy_names]
-    except boto.exception.BotoServerError as e:
-        if e.error_code == "NoSuchEntity":
-            # Role doesn't exist so it's safe to assume the policy doesn't either
-            module.exit_json(changed=False, msg="No such role, policy will be skipped.")
-        else:
-            module.fail_json(msg=e.message)
 
-    try:
-        matching_policies = []
-        for pol in current_policies:
-            if urllib.parse.unquote(iam.get_role_policy(name, pol).
-                                    get_role_policy_result.policy_document) == pdoc:
-                policy_match = True
-                matching_policies.append(pol)
+class Policy:
+
+    def __init__(self, client, name, policy_name, policy_document, policy_json, skip_duplicates, state, check_mode):
+        self.client = client
+        self.name = name
+        self.policy_name = policy_name
+        self.policy_document = policy_document
+        self.policy_json = policy_json
+        self.skip_duplicates = skip_duplicates
+        self.state = state
+        self.check_mode = check_mode
+        self.changed = False
+
+    @staticmethod
+    def _iam_type():
+        return ''
+
+    def _list(self, name):
+        return {}
+
+    def list(self):
+        return self._list(self.name).get('PolicyNames', [])
+
+    def _get(self, name, policy_name):
+        return '{}'
+
+    def get(self, policy_name):
+        return json.dumps(self._get(self.name, policy_name)['PolicyDocument'], sort_keys=True)
+
+    def _put(self, name, policy_name, policy_doc):
+        pass
 
-        if state == 'present':
-            # If policy document does not already exist (either it's changed
-            # or the policy is not present) or if we're not skipping dupes then
-            # make the put call.  Note that the put call does a create or update.
-            if not policy_match or (not skip and policy_name not in matching_policies):
-                changed = True
-                iam.put_role_policy(name, policy_name, pdoc)
-        elif state == 'absent':
+    def put(self, policy_doc):
+        if not self.check_mode:
+            self._put(self.name, self.policy_name, policy_doc)
+        self.changed = True
+
+    def _delete(self, name, policy_name):
+        pass
+
+    def delete(self):
+        self.changed = True
```

Comment:
This shouldn't be true for check mode if the policy has already been deleted. Maybe something along the lines of what you did for put()
```suggestion
        matching_policies = []
        policy_doc = self.get_policy_text()
        for pol in self.list():
            if self.get(pol) == policy_doc:
                matching_policies.append(pol)

        if self.policy_name in matching_policies:
            self.changed = True
```

Reply:
I thought this initially too, looks like @sbj-ss is relying on the `if e.response['Error']['Code'] == 'NoSuchEntity'` for that use-case

Reply:
That only works for non-check mode.